services/venice-server/src/main/java/com/linkedin/venice/listener/StorageReadRequestsHandler.java

```diff
@@ -506,7 +507,8 @@ private ReadResponse handleMultiGetRequest(MultiGetRouterRequestWrapper request)
     responseWrapper.setDatabaseLookupLatency(0);
     boolean isChunked = storageEngine.isChunked();
     for (MultiGetRouterRequestKeyV1 key: keys) {
-      int subPartitionId = getSubPartitionId(key.partitionId, topic, partitionerConfig, key.keyBytes.array());
+      int subPartitionId =
+          getSubPartitionId(key.partitionId, topic, partitionerConfig, ByteUtils.extractByteArray(key.keyBytes));
```

Comment:
Instead of `ByteUtils.extractByteArray`, can you change `getSubPartitionId` to accept a `ByteBuffer` instead? `ByteUtils.extractByteArray` will trigger a data copy every time

Reply:
This. `VenicePartitioner` already provides overload that accepts `ByteBuffer`

Reply:
Yeah, we should avoid data copy here. As this `getSubPartitionId` is also used by `singleGet`. For `singleGet`, the request comes with byte array naturally, so is it OK to convert that byte array to ByteBuffer to leverage new `getSubPartitionId` ?

Reply:
`keyBytes` is already `ByteBuffer` and can be passed directly to `VenicePartitioner`

Reply:
Using plain array is not safe and better be avoided.

Reply:
We can have two overloads of `getSubPartitionId`, one which takes a `byte[]` and the other a `ByteBuffer`, so there's no need to wrap in the single get case. Would be good to try to minimize code duplication across the two, however.

Reply:
I agree we should use ByteBuffer here.  My concern is only about the `getSubPartitionId` from `handleSingleGetRequest`, as the request in the method comes with byte array, do we need to wrap it into ByteBuffer?

Reply:
I think `ByteBuffer` is better, because there is no overhead comparing to plain array, and doing so will help to eliminate `getSubPartitionId(byte[])` overload in `StorageRequestHandler`

Reply:
Double checked the code: single-get don't use `ByteBuffer`, so we have to use both to avoid allocating new `ByteBuffer` object (bad for GC).